pkg/apis/pipeline/v1beta1/param_context.go

```diff
@@ -130,16 +134,26 @@ func getContextParams(ctx context.Context, overlays ...Param) []Param {
 		p := Param{
 			Name: ps.Name,
 		}
-		if ps.Type == ParamTypeString {
+		switch ps.Type {
+		case ParamTypeString:
 			p.Value = ArrayOrString{
 				Type:      ParamTypeString,
 				StringVal: fmt.Sprintf("$(params.%s)", ps.Name),
 			}
-		} else {
+		case ParamTypeArray:
 			p.Value = ArrayOrString{
 				Type:     ParamTypeArray,
 				ArrayVal: []string{fmt.Sprintf("$(params.%s[*])", ps.Name)},
 			}
+		default:
```

Comment:
> I think when the ParamSpec reaches here, the type should never be empty as webhook should have set the defaults.

Correct, mutating webhook will be able to infer the default type by its specification and set the type to one of the three types accordingly. And at this point here, the type has already been set to one of the three. Therefore, the default case will be object if the type is not `ParamTypeString` or `ParamTypeArray` and we don't need to throw error here. (In addition, as this function is used in `pkg/apis/pipeline/v1beta1/pipeline_defaults.go`, it seems mutating webhook does not throw errors generally.)

In the original implementation, the if-else statement also works like the default is array type instead of throwing error.

Reply:
> > In the original implementation, the if-else statement also works like the default is array type instead of throwing error. So I guess it makes more sense to follow the the original implementation.
> 
> I'm fine with either of the implementation. Both LGTM!

Thought we'd agreed on following the original implementation in which no error is thrown in default - `else` because I don't see mutating webhook (setdefaults) which throws errors.

Can you elaborate with an example about how to throw errors?

Reply:
Yes let's let other places to catch this error - at least don't set the value to be an object at default. Or am I looking at an old snapshot? You can just set the type to be ps.Type at default, and don't add anything for the value. This way some other module should catch this issue

Reply:
> At least don't set the value to be an object at default. Or am I looking at an old snapshot?

Only three possible types: string, array, object. And string & array have been caught in the first two cases and only object type is left. Did I miss some case?

I guess you're referring to unmarshal snippet in this [comment](https://github.com/tektoncd/pipeline/pull/4786#discussion_r863232810) where yongxuan requested default case is string for the purpose of result side. And that has been changed to default to string.

Reply:
> At least don't set the value to be an object at default. Or am I looking at an old snapshot?

Only three possible types: string, array, object. And string & array have been caught in the first two cases and only object type is left. Did I miss some case?

I guess you're referring to unmarshal snippet in this [comment](https://github.com/tektoncd/pipeline/pull/4786#discussion_r863232810) where yongxuan requested default case is string for the purpose of result side. And that has been changed to default to string.

Reply:
This should be:
```
case ParamTypeString:
...
case ParamTypeArray:
...
case ParamTypeObject:
...
default:
// set type to string
```
Ideally the default case should be throwing errors to catch this `ideally not should be happening` issue, but I agree that I don't see a good way to throw this error, but the cases should be separate.

Reply:
so in general I agree with @ywluogg that we should error out for unexpected scenarios but I think in this case this function is being called as part of the mutating webhook? If this is a general purpose function that is reused elsewhere - if so we could return an error from this function but handle it properly in the webhook (i.e. set the default). 

(I might be missing context here so let me know if I'm wrong!)

Reply:
> so in general I agree with @ywluogg that we should error out for unexpected scenarios but I think in this case this function is being called as part of the mutating Webhook?

Agreed! this is part of the mutating webhook.



> This should be:
> 
> ```
> case ParamTypeString:
> ...
> case ParamTypeArray:
> ...
> case ParamTypeObject:
> ...
> default:
> // set type to string
> ```
> 

So do you mean it'd better if we treat the `ParamTypeString ` case as the default (not throwing error)? If so, that sgtm!

Reply:
Yeah, because I don't think we want to make default to be objects, and this should be better aligned with https://github.com/tektoncd/community/blob/main/teps/0075-object-param-and-result-types.md#defaulting-to-string-types-for-values consistently